File: _experimental/_c/numbers/src/implementations/numbers_iterative.c
#include "../interfaces/numbers_iterative.h"

int fibonacci(int n) {
    return fibonacci_iter(n, 0, 1);
}

int factorial(int n) {
    return factorial_iter(n-1, n);
}

int sum_numbers(int n) {
    return sum_numbers_iter(n, 0);
}

int fibonacci_iter(int n, int acc2, int acc1) {
    if (n <= 1) {
        return n;
    } else if (n == 2) {
        return acc2 + acc1;
    } else {
        return fibonacci_iter(n - 1, acc1, acc1 + acc2);
    }
}

int factorial_iter(int n, int accumulator) {
    if (n == 1) {
        return accumulator;
    } else {
        return factorial_iter(n - 1, n * accumulator);
    }
}

int sum_numbers_iter(int n, int accumulator) {
    if (n == 0) {
        return accumulator;
    } else {
        return sum_numbers_iter(n - 1, n + accumulator);
    }
}

File: _experimental/_dart/numbers/lib/numbers_iterative.dart
class NumbersIterative {
  int fibonacci(int n) {
    return fibonacciIter(n, 0, 1);
  }

  int fibonacciIter(int n, int acc2, int acc1) {
    if (n <= 1) {
      return n;
    } else if (n == 2) {
      return acc2 + acc1;
    } else {
      return fibonacciIter(n - 1, acc1, acc1 + acc2);
    }
  }

  int factorial(int n) {
    return factorialIter(n - 1, n);
  }

  int factorialIter(int n, int accumulator) {
    if (n <= 1) {
      return accumulator;
    }
    return factorialIter(n - 1, n * accumulator);
  }

  int sumNumbers(int n) {
    return sumNumbersIter(n - 1, n);
  }

  int sumNumbersIter(int n, int accumulator) {
    if (n <= 0) {
      return accumulator;
    }
    return sumNumbersIter(n - 1, n + accumulator);
  }
}
